### PR TITLE
Update Custom C++ and CUDA Operators to use PyTorch stable API/ABI

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
## Description
Migrate `advanced_source/cpp_custom_ops.rst` to libtorch stable ABI/API. Accompanying change to [`extension-cpp`](https://github.com/pytorch/extension-cpp/pull/126)

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

cc @zou3519

### CODE REVIEW

File: advanced_source/cpp_custom_ops.rst [hunk 1]
```diff
@@ -380,8 +469,7 @@ three ways:
 Adding training (autograd) support for an operator
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 Use ``torch.library.register_autograd`` to add training support for an operator. Prefer
-this over directly using Python ``torch.autograd.Function`` or C++ ``torch::autograd::Function``;
-you must use those in a very specific way to avoid silent incorrectness (see
+this over directly using Python ``torch.autograd.Function`` (see
```
Comment (mikaylagawarecki) [inline advanced_source/cpp_custom_ops.rst:599]: remove reference to torch::autograd::Function since we don't support that in stable API/ABI

File: advanced_source/cpp_custom_ops.rst [hunk 2]
```diff
@@ -148,33 +158,67 @@ like so:
...
+the extension targets the LibTorch Stable ABI with a minimum supported PyTorch version of 2.10. The version format is:
+``[MAJ 1 byte][MIN 1 byte][PATCH 1 byte][ABI TAG 5 bytes]``, so 2.10.0 = ``0x020a000000000000``.
+
+See the section below for examples of code using the LibTorch Stable ABI.
+
 
 Defining the custom op and adding backend implementations
 ---------------------------------------------------------
-First, let's write a C++ function that computes ``mymuladd``:
+First, let's write a C++ function that computes ``mymuladd`` using the LibTorch Stable ABI:
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst]: Hmm same question here as on the other PR. Do we have enough in 2.10 to always recommend the libtorch stable APIs? If not, then we're going to want a section with libtorch stable apis and a section with the regular libtorch APIs.

Comment (mikaylagawarecki) [reply advanced_source/cpp_custom_ops.rst]: I updated extension-cpp to have both a libtorch stable and an unstable extension and refer the user to them at the top. The snippets in this PR are from the stable variant but I beefed up the section on "Reverting to the Non-Stable LibTorch API" and included 1 code snippet there, referring the user to the non stable extension in extension-cpp otherwise.

Does this sound good to you?

(btw the link check is failing because the extension-cpp pr needs to land first)

File: advanced_source/cpp_custom_ops.rst [hunk 3]
```diff
@@ -16,7 +16,7 @@ Custom C++ and CUDA Operators
     .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites
        :class-card: card-prerequisites
 
-       * PyTorch 2.4 or later
+       * PyTorch 2.10 or later (or PyTorch 2.4 or later if using the non-stable API)
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst]: "PyTorch 2.4 or later (or 2.10 or later if using the Stable ABI)"

File: advanced_source/cpp_custom_ops.rst [hunk 4]
```diff
@@ -62,12 +71,19 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
 
   setup(name="extension_cpp",
         ext_modules=[
-            cpp_extension.CppExtension(
+              cpp_extension.CppExtension(
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst]: indent looks wrong

File: advanced_source/cpp_custom_ops.rst [hunk 5]
```diff
@@ -555,28 +645,115 @@ When defining the operator, we must specify that it mutates the out Tensor in th
...
 
 .. note::
 
   Do not return any mutated Tensors as outputs of the operator as this will
   cause incompatibility with PyTorch subsystems like ``torch.compile``.
 
+.. _reverting-to-non-stable-api:
+
+Reverting to the Non-Stable LibTorch API
+----------------------------------------
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst]: nit: call it "Using the Non-ABI-Stable LibTorch API".
Some people just might not have upgraded to using latest PyTorch.

**Comment by mikaylagawarecki** [comment]
@AlannaBurke is this good to merge?

**Review by AlannaBurke — APPROVED** [review]
This looks good, the tabbed content looks really nice!
